### ISSUE TITLE
[DO NOT USE] Insecure / Security issues, reasons and an alternative

### ISSUE THREAD

**Issue by psibean** [valexandersaulys/tiny-csrf#1]
**You should immediately flag all current versions with a critical npm audit vulnerability and a github security advisory/report (from githubs securty tab) until fixed**

This code uses csurf as a reference for it's implementation, copying insecure practices and even introducing NEW insecure practices. csurf was shutdown and deprecated for a reason, because it was not secure. **This code is not secure:**

```
      if (
        csrfToken != undefined &&
        (req.query._csrf === csrfToken ||
          req.params._csrf === csrfToken ||
          req.body._csrf === csrfToken)
      ) 
```

You can watch this OWASP video to see why this is a problem: https://www.youtube.com/watch?v=2uvrGQEy8i4 @11:30 onwards

* The above code is insecure.
* The [OWASP CSRF Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html) recommends using a secure cookie, either signed, or hashing the token value before putting it into the cookie. Such an implementation should support rotating secrets.
* Default values should be SECURE. POST requests are not the only request that require protection, by default, `POST, PUT, DELETE` and `PATCH` requests should be protected.
* No Same-Site or httpOnly options

I've published [csrf-csrf](https://www.npmjs.com/package/csrf-csrf) which covers these cases and now supports both ESM and CommonJS natively as of v2.1.0

**Comment by psibean** [comment]
This should be fixed immediately by:

* Supporting signed cookies, or at least hashing the token before putting it into the cookie, but ideally you allow for both.
* Be explicit about WHERE the csrf token should be coming from on the request side. Ideally this should be inside a header that is prefixed with `Host__`  for additional security. But, you should also allow it to be customisable.
* Expose the sameSite and httpOnly options with reasonable defaults.
* If you don't hash the token value yourself, then set signed to true by default.
* Make it so all sensitive request types are protected by default, not just POST.

I recommend you refer to my csrf-csrf implementation if you wish to update yours.

**Comment by valexandersaulys** [comment]
Thanks for reaching out! Your suggestions have been implemented as of v1.1.

I appreciate letting me know about your library. The insights make me feel very enlightened. I've incorporated a few suggestions you stated.

My issue with a lot of CSRF libraries is that they are heavy to use. Having good security for a web app is essential. But the most secure web app is also a site that never gets built and never gets used.

This library is meant to be a drop-in. It's not meant to be the final word on CSRF related security. Perhaps that should be more clear in the README.

**Comment by psibean** [comment]
> Thanks for reaching out! Your suggestions have been implemented as of v1.1.

Thank you so much for taking my input so wholeheartedly. Seriously it's appreciated - I'm definitely aware I can be very aggressive some times, it's not the intention, it's just how I am. But, security issues are also quite serious.

You should raise a Github Security Advisory from the github security tab specifying the impacted versions. This way the vulnerability will show up on an npm audit. This means users will be informed they need to update when they do an install.

Mark the issue as Critical

Going over the changes you have made, I believe they are sufficient. Although you should definitely specify in the readme that the code does not follow all of the best practices and recommendations from OWASP.
